### PR TITLE
Implement cw721-base nft contract

### LINKED ISSUE #44: Implement base NFT contract
Builds on #43 

Implement a simple contract that implements the NFT spec (like cw20-base does). This is means to be used stand-alone or further customized.

Not not implement enumeration, or at least enumeration by owner, as that requires secondary indexes to do cleanly: https://github.com/CosmWasm/cosmwasm/issues/468

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #44

Most logic writen.

TODO:
- [x] Send proper message on `SendNftToken`
- [x] Add unit tests
- [x] Write a better README

### CODE REVIEW

**Comment by ethanfrey** [comment]
@maurolacy This is roughly done for the main code path, only basic unit tests.
It also could use a good review to point out any issues in the design.

**Comment by maurolacy** [comment]
OK, I'll review it tomorrow morning.

File: contracts/cw721-base/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+}
+
+pub fn handle_send_nft<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    contract: HumanAddr,
+    token_id: String,
+    _msg: Option<Binary>,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
```
Comment (maurolacy) [inline contracts/cw721-base/src/contract.rs:130]: This is just `handle_transfer_nft()` (plus the send msg), and can probably be simplified as such.

File: contracts/cw721-base/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_revoke<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    spender: HumanAddr,
+    token_id: String,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
```
Comment (maurolacy) [inline contracts/cw721-base/src/contract.rs:186]: Not sure it's worth it, but this logic is shared between `approve` and `revoke`. In fact, `approve` is just `revoke` plus adding a new spender.

**Review by maurolacy — COMMENTED** [review]
Looks pretty good to me. Besides what I think is a serious error (`token_id` overwrite in minter), logic looks good.

Will continue reviewing the tests later.

File: contracts/cw721-base/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+
+    // create the token
+    let token = TokenInfo {
+        owner: deps.api.canonical_address(&owner)?,
+        approvals: vec![],
+        name,
+        description: description.unwrap_or_default(),
+        image,
+    };
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply contracts/cw721-base/src/contract.rs:97]: This is very true. Good catch and major bug

**Comment by ethanfrey** [comment]
Addressed the major issues raised - good points, thank you.

**Comment by maurolacy** [comment]
I've made some changes directly on your branch. Please review them.

File: contracts/cw721-base/src/contract.rs [hunk 4]
```diff
@@ -397,7 +397,7 @@ fn humanize_approvals<A: Api>(api: A, info: &TokenInfo) -> StdResult<Vec<cw721::
 fn humanize_approval<A: Api>(api: A, approval: &Approval) -> StdResult<cw721::Approval> {
     Ok(cw721::Approval {
         spender: api.human_address(&approval.spender)?,
-        expires: approval.expires.clone(),
```
Comment (ethanfrey) [inline contracts/cw721-base/src/contract.rs]: Yup, this is a nice fix building on cw3 pr.

File: contracts/cw721-base/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    _env: Env,
+    msg: InitMsg,
+) -> StdResult<InitResponse> {
+    set_contract_version(&mut deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let info = ContractInfoResponse {
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply contracts/cw721-base/src/contract.rs:29]: Yeah, I figure it is 1:1 what we return (and only for returning that info) no need to define a new type.

But in general, nice to keep state and api separate.

**Pull request merged**